src/include/miopen/sparsesoftmaxcrossentropywithlogits/problem_description.hpp

```diff
@@ -0,0 +1,157 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/miopen.h>
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace sparsesoftmaxcrossentropywithlogits {
+
+struct FwdProblemDescription : ProblemDescriptionBase
+{
+    FwdProblemDescription(const TensorDescriptor& inputDesc_,
+                          const TensorDescriptor& targetDesc_,
+                          const TensorDescriptor& outputDesc_,
+                          const TensorDescriptor& backpropDesc_)
+        : inputDesc(inputDesc_),
+          targetDesc(targetDesc_),
+          outputDesc(outputDesc_),
+          backpropDesc(backpropDesc_)
+    {
+        IsSameType();
+        IsValidType();
+        IsValidDims();
+    }
+
+    const TensorDescriptor& GetInputDesc() const { return inputDesc; }
+    const TensorDescriptor& GetTargetDesc() const { return targetDesc; }
+    const TensorDescriptor& GetOutputDesc() const { return outputDesc; }
+    const TensorDescriptor& GetBackpropDesc() const { return backpropDesc; }
+
+    bool IsSameType() const
+    {
+        if(inputDesc.GetType() != outputDesc.GetType() ||
+           inputDesc.GetType() != backpropDesc.GetType())
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "SparseSoftmaxCrossEntropyWithLogitsForward: Data types do not match.");
+        return true;
+    }
+
+    bool IsValidType() const
+    {
+        if(targetDesc.GetType() != miopenInt32)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "SparseSoftmaxCrossEntropyWithLogitsForward: target tensor must be int32, "
+                         "unsign int32, or unsign int64.");
+        }
+        return true;
+    }
+
+    bool IsValidDims() const
+    {
+        if(inputDesc.GetNumDims() != 2 || targetDesc.GetNumDims() != 1 ||
+           outputDesc.GetNumDims() != 1 || backpropDesc.GetNumDims() != 2)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "SparseSoftmaxCrossEntropyWithLogitsForward: Tensor sizes do not match.");
+        }
+        return true;
+    }
+
+    bool IsAllContiguous() const
+    {
+        return inputDesc.IsContiguous() && targetDesc.IsContiguous() && outputDesc.IsContiguous() &&
+               backpropDesc.IsContiguous();
+    }
```

Comment:
Contiguous kernel also works if all tensor are `packed` and have same `stride`. This condition should be changed or replaced by `allPacked` and `SameStride` constraints.

Reply:
This shouldn't be change because they can't all have same stride because input and backprop are 2d while target and output are 1d tensor.

Reply:
agree